.github/workflows/elixir.yml

```diff
@@ -34,44 +50,51 @@ jobs:
     - name: Restore deps cache
       uses: actions/cache@v2
       with:
-        path: deps
-        key: deps-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}
+        path: |
+          deps
+          _build
+        key: ${{ runner.os }}-deps-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}-git-${{ github.sha }}
```

Comment:
Seems like the `sha` is sub-ideal here, it's _always_ different.

I wonder if we should `hashFiles('deps/*')`

Reply:
`t's always different.`
And that's the main point here.

When `action/cache` is looking up for cache, it will firstly try `key`, then if nothing was found I'll try all the items from `restore-keys` until something will be found.
Also if it downloaded cache by primary key (`key`), it will not upload the updated files to cache storage.

If you have only `key` without `restore-keys`, `action/cache` will download it but will never update. In this case every update of mix.lock will lead to generating new primary key, and thus using completely new cache. But it can be a very long operation for large projects.

If you have `key: {{otp}}-{{elixir}}-{{hash(mix.lock)}}` and `restore-keys: {{otp}}-{{elixir}}-`, it will work faster because even you've upgraded mix.lock, and file hash was changed, 'action/cache` will download the latest cache item which correspond to the same OTP and Elixir versions. Mix will left intact dependencies which were not changed, and will update only those which were updated.

But sometimes there are steps which change your cached files but not changing the mix.lock hash. It can be a `dialyzer` run (it stores `.plts` files in `deps` subfolder), or some other tool which saves data to `deps` or `_build` folders.
In this case `action/cache` decides that because cache was fetched by primary key (mix.lock wasn't changed), there is no need to update the existing cache in the cache storage. But without updating this cache dialyzer and other tools like that will rebuild their files every time, and sometimes it takes a lot of time.

So you basically have a cache, but because it works not as expected, it will increase time of every next CI job run. Until `mix.lock` file will be updated which leads to updating the cache.

That's why I propose to add `github.sha` into the primary key here - it will always be new, so `action/cache` step will update cache every time. And because there are items in `restore-keys`, this cache will still be downloaded on next commits, so it will not be disabled as you may expect.

You cannot put `hash(**/_build)` into the key because this directory does not exist in the commit.